src/lib/mina_base/test/zkapp_command/call_forest.ml

```diff
@@ -0,0 +1,320 @@
+open Core_kernel
+open Signature_lib
+open Mina_base
+open Mina_base.Zkapp_command.Call_forest
+
+(* TODO: move generators to the library once the code isn't actively being
+     worked on *)
+module Tree = struct
+  include Tree
+
+  module Stable = struct
+    include Stable
+
+    module V1 = struct
+      include V1
+
+      let quickcheck_generator account_update_gen account_update_digest_gen
+          digest_gen =
+        let open Quickcheck.Generator.Let_syntax in
+        Quickcheck.Generator.fixed_point (fun self ->
+            let%bind calls_length =
+              Quickcheck.Generator.small_non_negative_int
+            in
+            let%map account_update = account_update_gen
+            and account_update_digest = account_update_digest_gen
+            and digest = digest_gen
+            and calls =
+              Quickcheck.Generator.list_with_length calls_length
+                (With_stack_hash.quickcheck_generator self digest_gen)
+            in
+            { account_update; account_update_digest; calls } )
+    end
+  end
+end
+
+module Shape = struct
+  include Shape
+
+  let rec sexp_of_t = function
+    | Node l ->
+        let rec sexp_of_t_aux (i, t) =
+          Sexp.List [ Sexp.Atom (Int.to_string i); sexp_of_t t ]
+        in
+        Sexp.List (List.map ~f:sexp_of_t_aux l)
+
+  let rec compare (Node l_x) (Node l_y) =
+    List.compare
+      (fun (i_x, s_x) (i_y, s_y) ->
+        match Int.compare i_x i_y with 0 -> compare s_x s_y | n -> n )
+      l_x l_y
+end
+
+let quickcheck_generator account_update_gen account_update_digest_gen digest_gen
+    =
+  let open Quickcheck.Generator.Let_syntax in
+  Quickcheck.Generator.list
+    (With_stack_hash.Stable.V1.quickcheck_generator
+       (Tree.Stable.V1.quickcheck_generator account_update_gen
+          account_update_digest_gen digest_gen )
+       digest_gen )
+
+module Tree_test = struct
+  let tree i calls =
+    { Tree.calls; account_update = i; account_update_digest = () }
+
+  let node i calls = { With_stack_hash.elt = tree i calls; stack_hash = () }
+
+  let%test_unit "fold_forest" =
+    [%test_result: int]
+      (Tree.fold_forest [] ~f:(fun _ _ -> 0) ~init:1)
+      ~expect:1 ;
+    [%test_result: int]
+      (Tree.fold_forest
+         [ node 0 [ node 1 [ node 1 [] ] ]; node 2 [ node 3 [] ] ]
+         ~f:(fun acc x -> acc + x)
+         ~init:0 )
+      ~expect:7
+
+  let%test_unit "fold_forest2" =
+    [%test_result: int]
+      (Tree.fold_forest2_exn [] [] ~f:(fun _ _ _ -> 0) ~init:1)
+      ~expect:1 ;
+    [%test_result: int]
+      (Tree.fold_forest2_exn
+         [ node 0 [ node 1 [ node 2 [] ] ]; node 3 [ node 4 [] ] ]
+         [ node 5 [ node 6 [ node 7 [] ] ]; node 8 [ node 9 [] ] ]
+         ~f:(fun acc x y -> acc + x + y)
+         ~init:0 )
+      ~expect:45 ;
+    try
```

Comment:
The with-handler could be catching an exception from the fold, or from the `assert false`. If the fold does not raise, the `assert false` will raise, and this part succeeds. If the fold raises, this part succeeds.

I'm guessing the fold is meant to raise.

Maybe put this part in a separate test that uses a boolean result, put `false` after the fold (not `assert false`), and the with-handler returns `true`.